arangod/Aql/OptimizerRules.cpp

```diff
@@ -6929,28 +6933,41 @@ void arangodb::aql::geoIndexRule(Optimizer* opt, std::unique_ptr<ExecutionPlan>
   plan->findNodesOfType(nodes, EN::ENUMERATE_COLLECTION, true);
   for (ExecutionNode* node : nodes) {
     GeoIndexInfo info;
+    info.collectionNodeToReplace = node;
```

Comment:
Because if the target collection is not set here, it will only work for queries with one collection/geo index.
But consider
```
FOR doc1 IN geo1
  FOR doc2 IN geo2
    FOR doc3 IN geo3
      FILTER DISTANCE(doc1.lat, doc1.lon, ...) < ... && 
             DISTANCE(doc2.lat, doc2.lon, ...) < ... && 
             DISTANCE(doc3.lat, doc3.lon, ...) < ... && 
```
In this case the optimizer may have permuted the loops, and the FILTER condition contains a few nested logical ANDs.
It is therefore very unclear which FILTER condition we will encounter first, and if it will match the EnumerateCollectionNode we are currently looking at.

The old rule simply hasn't expected such queries and never worked with it.
Now, whenever we want to find the filter conditions for an EnumerateCollectionNode, we first set the node in the `info` struct. That alone does no harm, as it is not enough to make the EnumerateCollectionNode be replaced with an IndexNode. The `info` struct is only complete if its operator() returns true, and that is only when `return collectionNodeToReplace != nullptr && collectionNodeOutVar && collection && index && valid;`

And the benefit is that setting the EnumerateCollectionNode in the `info` struct prevents that other, non-related filters (for other geo indexes) are picked up on enter themselves into the `info` struct, which however are not related to the current `EnumerateCollectionNode`. That can prevent the correct filters from being picked up. It is all order-dependent, but for complex queries the order may be changed around by the optimizer a lot.

I verified it, and when not setting the node in the `info` struct beforehand, it does not work reliably with multiple collections and filters. And when setting it, it just works.